src/main/java/io/strimzi/kafka/bridge/http/HttpUtils.java

```diff
@@ -17,15 +17,19 @@
     private static final Logger log = LoggerFactory.getLogger(HttpUtils.class);
 
     public static void sendResponse(RoutingContext routingContext, int statusCode, String contentType, Buffer body) {
-        routingContext.response().setStatusCode(statusCode);
-        if (body != null) {
-            log.debug("[{}] Response: body = {}", routingContext.get("request-id"), Json.decodeValue(body));
-            routingContext.response().putHeader(HttpHeaderNames.CONTENT_TYPE, contentType);
-            routingContext.response().putHeader(HttpHeaderNames.CONTENT_LENGTH, String.valueOf(body.length()));
-            routingContext.response().write(body);
+        if (!routingContext.response().closed() && !routingContext.response().ended()) {
+            routingContext.response().setStatusCode(statusCode);
+            if (body != null) {
+                log.debug("[{}] Response: body = {}", routingContext.get("request-id"), Json.decodeValue(body));
+                routingContext.response().putHeader(HttpHeaderNames.CONTENT_TYPE, contentType);
+                routingContext.response().putHeader(HttpHeaderNames.CONTENT_LENGTH, String.valueOf(body.length()));
+                routingContext.response().write(body);
+            }
+            routingContext.response().end();
+            logResponse(routingContext);
+        } else {
+            log.debug("[{}] Response: already closed/ended!", routingContext.get("request-id").toString());
```

Comment:
As explained in the related issue, the scenario is the following ...

If I kill the application I am using for tests, the shutdown hook undeploys the consumer verticle which stops the timer about sending poll requests and calls the delete consumer. At this point in the shutdown, I wait that the delete is done properly with a timeout: doing more long poll requests in the above scenarios it can happen that the delete is postponed so the application just ends for timeout on the latch. When on the bridge the poll will come back, it will try to write a response on a closed/ended socket. Because the consumer isn't deleted properly, the bridge has no knowledge that the application is no more interested in the poll responses.

Reply:
I think you're conflating two different things in the code currently:

* The response is ended. It can only be the bridge which ended it. In which case trying to send a response to an already ended response must indicate a bug in the bridge and logging at debug is not the right thing to do.
* The TCP connection is closed. Presumably it could have been closed by the bridge, or on the client side.
    * If the connection was closed by the bridge (because the verticle is being undeployed, for example) then I would argue that it is again a bug in the bridge: when the bridge closes a TCP connection it should be waiting for outstanding responses to finish, or at least sending responses for any outstanding requests. However this is inherently racy (since the client can keep making requests while we're trying to send responses for existing outstanding requests), so maybe we need to think more about how to handle this corner case.
    * If the connection was closed by the client (which is, I think, the situation in your tests), then the debug logging stmt is fine, because there's nothing else the bridge can do.

Reply:
> The response is ended. It can only be the bridge which ended it.

Is this assumption based on the fact that there is an `end()` method on the response. Maybe it's considered ended even when the remote client close the connection or maybe not. Btw the condition I have today is 
```
if (!routingContext.response().closed() && !routingContext.response().ended())
```

so maybe the trace I see is related to the fact that the response is closed and it happens even when remote client closes the connection.
Let me to be more precise about the conditions.

> If the connection was closed by the bridge (because the verticle is being undeployed, for example)

I am not closing the bridge, why it should close the connection to the client?

Reply:
Just adding that, using Wireshark, the FYN is sent by the client and that then the `Response: already closed/ended!` log is raised so I guess that it's because the response is closed not ended ... let me check ;)

Reply:
@tombentley we could have reached a good point where "no bugs" seem to be in the bridge. After my tests and logging response closed/ended more precisely ... I see that in my scenarios ...
```
2019-06-18 16:54:25,968 [     vert.x-eventloop-thread-3] HttpUtils                      DEBUG [1907003446] Response: already closed [true] - ended [false]
2019-06-18 16:54:30,969 [     vert.x-eventloop-thread-3] HttpUtils                      DEBUG [1036567481] Response: already closed [true] - ended [false]
2019-06-18 16:54:35,970 [     vert.x-eventloop-thread-3] HttpUtils                      DEBUG [500698412] Response: already closed [true] - ended [false]
2019-06-18 16:54:40,971 [     vert.x-eventloop-thread-3] HttpUtils                      DEBUG [1200619310] Response: already closed [true] - ended [false]
2019-06-18 16:54:45,972 [     vert.x-eventloop-thread-3] HttpUtils                      DEBUG [1652686025] Response: already closed [true] - ended [false]
```

The log is due to connection closed by the clients and never response ended as you mentioned by the bridge.
If it's fine with you now we should decide how to log this info. Anyway in this early age of the bridge having something logged could be really useful to study different corner cases ... maybe coming back to warn?

Reply:
> Is this assumption based on the fact that there is an end() method on the response.

That and the javadoc on `response.end()` and `response.ended()` makes it pretty clear to me that the ended state is exclusive of connection closed state.

>  Anyway in this early age of the bridge having something logged could be really useful to study different corner cases ... maybe coming back to warn?

How to you propose to get the logs for those different corner cases? If you can't get those logs then this extra logging is worthless. It would be better to log at WARN in those cases where you think it would really be a bug in the bridge. That way an absence of complaints about those warn messages is a little move convincing as evidence that the code is correct.

Reply:
For starting I would propose to log as WARN just when the response is ended which could be a bridge bug. Wdyt?

Reply:
That sounds good to me.